### ISSUE TITLE
DetailsList Accessibility Issues

### ISSUE THREAD

**Issue by JasonGore** [OfficeDev/office-ui-fabric-react#5115]
### Bug Report

- __Package version(s)__: OUFR v6.5.0
- __Browser and OS versions__: n/a

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? Yes

Requested priority: Normal

Products/sites affected: (if applicable)

#### Describe the issue:
DetailsList does not follow ARIA spec in a few areas and does not observe the `ariaLabel` prop.

#### Actual behavior:

1. Per [ARIA navigation spec](https://www.w3.org/TR/wai-aria-practices-1.1/#grid), components with "grid" role as a whole should only have one tab stop. DetailsList is tab stopping on both the header and the list.
1. `ariaLabel` prop does not appear to be respected by `DetailsList` and does not appear in the DOM hierarchy.
1. The `aria-labelledby` attribute should be on the component with role `columnheader`, not in a component lower in the hierarchy. (https://www.w3.org/TR/wai-aria-1.1/#columnheader)

#### Expected behavior:

Respect ARIA spec and props documentation.

### If applicable, please provide a codepen repro:

https://codepen.io/jasongore/pen/bzxNBX

**Comment by ThomasMichon** [comment]
For (1), we'll need Design to chime in. @betrue-final-final do we want to change whether or not the header gets its own Tab stop?

**Comment by dzearing** [comment]
It seems to me that the tabbing behavior is by design, but the roles might not be.

Could we stick `role="grid"` on the surface area? (excluding the column headers?)

I'd like to either come up with an action plan that makes sense here, or close as by design.

**Comment by KevinTCoughlin** [comment]
@dzearing original issue answers in-line:

> * Per [ARIA navigation spec](https://www.w3.org/TR/wai-aria-practices-1.1/#grid), components with "grid" role as a whole should only have one tab stop. DetailsList is tab stopping on both the header and the list.

Using [W3's "Data Grid" Example 2](https://www.w3.org/TR/wai-aria-practices-1.1/examples/grid/dataGrids.html#ex2_label) for comparison, it _does_ look like we should only have **one** tab stop. In the example, header, rows, and cells are navigated using arrow keys once the root element receives focus.

The above behavior change will affect document libraries across OneDrive & SharePoint.

> * `ariaLabel` prop does not appear to be respected by `DetailsList` and does not appear in the DOM hierarchy.

DetailsList supports `ariaLabel` on the root element and a number of inner components, see [DetailsList with ariaLabel on root element](https://codepen.io/kevintcoughlin/pen/bzmGVo). @JasonGore do you remember what is meant by the above?

> * The `aria-labelledby` attribute should be on the component with role `columnheader`, not in a component lower in the hierarchy. (https://www.w3.org/TR/wai-aria-1.1/#columnheader)

Also unsure about the above. Inspect https://codepen.io/kevintcoughlin/pen/bzmGVo and grep for `aria-labelledby`. This looks like a non-issue, but my understanding may be incorrect.

![image](https://user-images.githubusercontent.com/706967/52813915-b93f6500-304f-11e9-9b29-442475ce88e1.png)

---

Given the above, we need to decide if we want to support the first scenario and come up with a plan which includes the product teams that will be affected / get their sign-off cc: @aditima @natalieethell.

From my experience, the accessibility issue re: row selection state is more pressing #4240. That issue requires a dedicated dev to debug, fix, and _more importantly_ introduce guards for regression.

**Comment by JasonGore** [comment]
I created this issue in June of 2018, so it's entirely possible these issues have been fixed since. I agree that item 2 and 3 no longer appear to be issues, but item 1 is in the ARIA navigation spec and so would consider that something we should look at fixing.

**Comment by KevinTCoughlin** [comment]
@dzearing @aditima thoughts on editing this issue to capture the ask of overhauling tab behavior? This would be feature+ work given the collaboration required with partners (I think).

I will note that I haven't seen this accessibility ask elsewhere, so I'm unsure of its priority.

**Comment by JasonGore** [comment]
Specifically in the [ARIA navigation spec](https://www.w3.org/TR/wai-aria-practices-1.1/#grid) (with added highlighting):

> The grid pattern can be used to group a set of interactive elements, such as links, buttons, or checkboxes. **Since only one element in the entire grid is included in the tab sequence, grouping with a grid can dramatically reduce the number of tab stops on a page**. This is especially valuable if scrolling through a list of elements dynamically loads more of those elements from a large data set, such as in a continuous list of suggested products on a shopping site. **If elements in a list like this were in the tab sequence, keyboard users are effectively trapped in the list.** 

The concern with limiting tab stops seems to be more about having an indefinite number of them as opposed to requiring it to be only 1. If our case is that we have 2 and it's deterministic and, in combination with the lack of asks as Kevin mentioned, maybe this isn't necessarily a must-fix.

**Comment by dzearing** [comment]
I don't really think DetailsList is a data grid. Excel spreadsheets are data grids, while the DetailsList is more like a data list. You down arrow and select a whole row. You can, optionally, drill into the cells of a row, but overall, moving vertically means moving between rows, and not cells. In a grid, cells are the important piece of data, rather than rows.

But even then, I strongly strongly disagree that the column headers should share the focuszone with the cells in a list OR a grid scenario. Even in Excel, cell headers have a lot of power, and tabbing between the content and the chrome makes a ton of sense for not losing context while trying to execute an action.

It's a similar scenario to why command bars and list items aren't in the same zone. Tabbing around a ui should jump from concern to concern. Cell management is a separate concern from column/row management.

**Comment by dzearing** [comment]
@JasonGore I think it's a won't fix...

**Comment by JasonGore** [comment]
I agree with everything you said, but regardless of what DetailsList is or does, it has elected to use ARIA role of `grid` which has its own requirements. Even the ARIA grid examples with headers treat them as the same focuszone.

But again, I think this is a blurry line based on the quotations I pasted above. I think we're following the spirit of ARIA `grid`, even if our header rows behave slightly different from ARIA's reference example.

**Comment by cschlechty** [comment]
I'm inclined to close this too. Note that the authoring practices are not proscriptive, but are W3's suggestions on how to implement the spec. if we vary a little, that should be fine

**Comment by cschlechty** [comment]
Per the docs https://www.w3.org/TR/wai-aria-practices-1.1/#intro

> 1. Introduction
> This section is informative.
> 
> WAI-ARIA Authoring Practices is a guide for understanding how to use WAI-ARIA 1.1 to create an accessible Rich Internet Application. It provides guidance on the appropriate application of WAI-ARIA, describes recommended WAI-ARIA usage patterns, and explains concepts behind them.
